### PR TITLE
CORDA-3880: Streamline re-connection logic in RPCClientProxyHandler

### PR DESCRIPTION
**Pull request by vkolomeyko**
Please see [Jira](https://r3-cev.atlassian.net/browse/CORDA-3880) for underlying reason for the change.

Also includes some minor tweaks to the whole class.

### CODE REVIEW

File: client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt [hunk 1]
```diff
@@ -76,10 +76,10 @@ import kotlin.reflect.jvm.javaMethod
  * forwarded to the [UnicastSubject]. Note that the observations themselves may contain further [Observable]s, which are
  * handled in the same way.
  *
- * To do the above we take advantage of Kryo's datastructure traversal. When the client is deserialising a message from
- * the server that may contain Observables it is supplied with an [ObservableContext] that exposes the map used to demux
- * the observations. When an [Observable] is encountered during traversal a new [UnicastSubject] is added to the map and
- * we carry on. Each observation later contains the corresponding Observable ID, and we just forward that to the
+ * To do the above we take advantage of serialisation data structure traversal. When the client is deserialising a message from
```
Comment (dimosr) [inline client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt:79]: Thanks for refreshing the docs.

File: client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt [hunk 2]
```diff
@@ -528,36 +530,36 @@ class RPCClientProxyHandler(
...
+        var reconnectAttempt = 1
         var retryInterval = rpcConfiguration.connectionRetryInterval
         val maxRetryInterval = rpcConfiguration.connectionMaxRetryInterval
 
-        var transportIterator = serverLocator.staticTransportConfigurations.iterator()
-        while (transportIterator.hasNext() && reconnectAttempts != 0) {
-            val transport = transportIterator.next()
-            if (!transportIterator.hasNext())
-                transportIterator = serverLocator.staticTransportConfigurations.iterator()
+        while (if (maxReconnectCount > 0) reconnectAttempt < maxReconnectCount else true) {
```
Comment (dimosr) [inline client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: having an if conditional inside a while statement makes it a bit hard to parse mentally.
It'd be better to have a private helper method, smth like `while(shouldRetry(reconnectAttempt, maxReconnectCount))`

Also, shouldn't the logic be instead:
`if (maxReconnectCount >= 0) reconnectAttempt <= maxReconnectCount else true`?

File: client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt [hunk 3]
```diff
@@ -528,36 +530,36 @@ class RPCClientProxyHandler(
...
+            val transport = serverLocator.staticTransportConfigurations.let { it[reconnectAttempt % it.size] }
```
Comment (dimosr) [inline client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: minor: maybe switch to `it[(reconnectAttempt-1) % it.size]`, so that we try addresses in the order specified by the user.

File: client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt [hunk 4]
```diff
@@ -566,8 +568,12 @@ class RPCClientProxyHandler(
             break
         }
 
-        if (reconnectAttempts == 0 || sessionFactory == null)
-            log.error("Could not reconnect to the RPC server.")
+        val maxReconnectReached = maxReconnectCount > 0 && reconnectAttempt >= maxReconnectCount
```
Comment (dimosr) [inline client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: similar to the above, I'd think this should be:
`maxReconnectCount >= 0 && reconnectAttempt > maxReconnectCount`

File: client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt [hunk 5]
```diff
@@ -566,8 +568,12 @@ class RPCClientProxyHandler(
...
+        if (maxReconnectReached || sessionFactory == null) {
+            val errMessage = "Could not reconnect to the RPC server after trying $reconnectAttempt times. Session factory was " +
+                    if (sessionFactory == null) "NOT " else "" + "present"
```
Comment (dimosr) [inline client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: what is the scenario in which `sessionFactory` would be null? I am afraid this log will be quite cryptic - is there a way to provide something more useful?

**Review by dimosr — CHANGES_REQUESTED** [review]

File: client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt (hunk 1 above)
Comment (vkolomeyko) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt:79]: 👍

File: client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt (hunk 2 above)
Comment (vkolomeyko) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: Thanks @dimosr 
I guess it is down to the contract for `maxReconnectCount`, which is not written anywhere. From the code when `maxReconnectCount` is `-1` then this means infinite number of reconnects. But what about when `maxReconnectCount == 0` ?
Since it is not written, my assumption was that  `maxReconnectCount = 0` also means infinite number of re-connects.

I take your point about helper function and will make a change accordingly.

Comment (dimosr) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: yeah, I also had a look at the docs, it's not completely clear, indeed. I assumed that any non-negative values would have the natural meaning, so 0 would mean no retries.

I wouldn't be against the current logic assuming we'll clarify the docs to make clear the special meaning of 0. Another question I'd have in this case is is there any way to remove retries if we take the option of setting it to zero away?

File: client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt (hunk 3 above)
Comment (vkolomeyko) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: Sure - makes sense not to skip `0` endpoint on first re-connect.

File: client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt (hunk 4 above)
Comment (vkolomeyko) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: Sure - will change.

File: client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt (hunk 5 above)
Comment (vkolomeyko) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: Looking deeper into the code, I think `sessionFactory` would be `null` if it was never possible to establish connection with any of the endpoints specified. I.e. method `createSessionFactory` was always ended-up throwing an exception.

I guess I will re-phrase the error message to highlight this fact.

File: client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt (hunk 2 above)
Comment (vkolomeyko) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: Sure - I will update the docs to be explicit.

> is there any way to remove retries

I guess there is a little bit of contradiction here. If the user providing a pool of endpoints of size greater than 1, why would such a user ever ask for no retries to happen?

Comment (dimosr) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: that's a good point, not sure. Maybe if someone wants to have all addresses tried on the first connection attempt, while at the same time failing fast and avoiding any retries on transient issues?

Comment (vkolomeyko) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: Yes, I agree - but this seem to be a very unusual type of requirement.

Comment (vkolomeyko) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: And I have updated the KDocs on `maxReconnectAttempts` to explicty say: `Any non-positive value would mean that there will be an infinite number of reconnect attempts.`

Comment (dimosr) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: Given this is essentially a refactoring change, I'd prefer to keep the old behaviour since there might be customers relying on it already.

Comment (vkolomeyko) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: I still think we have to explicitly state what is going to happen when a negative or zero value is passed into `maxReconnectAttempts`. Especially given that in our documentation we say: 
"See `CordaRPCClientConfiguration.maxReconnectAttempts` for adjusting the number of retries."

I doubt any of customers put a value of `0` to `maxReconnectAttempts`, but @dimosr do you think it would be correct to change the Kdocs to say `Any negative value would mean that there will be an infinite number of reconnect attempts.` and change the code accordingly?

Comment (dimosr) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: In cases like this, I usually prefer to be on the safe side and preserve existing behaviour (just in case someone relies on it) unless of course there's a clear requirement to change it. So, I'd do the following here:
* Adjust the kdocs to make clear what's the contract for negative values, i.e. infinite retries as you suggested. I believe that's already what the code does, so no code changes needed there.
* Change the code, so that it does not perform retries when `maxReconnectAttempts = 0`, similar to the existing behaviour. On the kdocs side, either call this out - the fact that zero means no retries - or not mention anything at all. I think this is the natural meaning of this value, so it shouldn't be a big problem.

Comment (vkolomeyko) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: OK, I will update the KDoc and change the code such that we do not re-connect if `maxReconnectAttempts = 0`.

Comment (vkolomeyko) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt]: @dimosr - I have pushed more changes as https://github.com/corda/corda/pull/6426/commits/cee7e796853b62f49abec0e681e8ce9f19488d24
Could you please re-review when you have time?

**Review by dimosr — APPROVED** [review]
LGTM

**Review by mnesbit — APPROVED** [review]
OK, but please check with Tamas
